ruby/neo4j/driver/internal/async/inbound/inbound_message_dispatcher.rb

```diff
@@ -0,0 +1,186 @@
+
+            handler = handlers.first
+            if handler.nil?
+              rails Neo4j::Driver::Exceptions::IllegalStateException.new("No handler exists to handle RECORD message with fields #{fields}")
+            end
+
+            handler.on_record(fields)
+          end
+
+          def handle_failure_message(code, message)
+            log.debug("S: FAILURE #{code}, '#{message}'")
+            current_error = Util::ErrorUtil.new_neo4j_error(code, message)
+
+            # we should not continue using channel after a fatal error
+            # fire error event back to the pipeline and avoid sending RESET
+            return channel.pipeline.fire_exception_caught(current_error) if Util::ErrorUtil.is_fatal?(current_error)
+
+            if current_error.kind_of?(Neo4j::Driver::Exceptions::AuthorizationExpiredException)
+              Connection::ChannelAttributes.authorization_state_listener(channel).on_expired(current_error, channel)
+            else
+              # write a RESET to "acknowledge" the failure
+              enqueue(ResetResponseHandler.new(self))
+              channel.write_and_flush(RESET, channel.void_promise)
+            end
+
+            invoke_before_last_handler_hook(HandlerHook::FAILURE)
+            handler = remove_handler
+            handler.on_failure(current_error)
+          end
+
+          def handle_ignored_message
+            log.debug('S: IGNORED')
+            handler = remove_handler
+
+            if !current_error.nil?
+              error = current_error
+            else
+              log.warn("Received IGNORED message for handler #{handler} but error is missing and RESET is not in progress. Current handlers #{handlers}")
+              error = Neo4j::Driver::Exceptions::ClientException('Database ignored the request')
+            end
+
+            handler.on_failure(error)
+          end
+
+          def handle_channel_inactive(cause)
+            # report issue if the connection has not been terminated as a result of a graceful shutdown request from its
+            # parent pool
+            if !gracefully_closed
+              handle_channel_error(cause)
+            else
+              channel.close
+            end
+          end
+
+          def handle_channel_error(error)
+            if !current_error.nil?
+              # we already have an error, this new error probably is caused by the existing one, thus we chain the new error on this current error
+              Util::ErrorUtil.add_suppressed(current_error, error)
+            else
+              current_error = error
+            end
+
+            fatal_error_occurred = true
+
+            while !handlers.empty?
+              handler = remove_handler
+              handler.on_failure(current_error)
+            end
+
+            error_log.trace_or_debug('Closing channel because of a failure', error)
+            channel.close
+          end
+
+          def clear_current_error
+            current_error = nil
```

Comment:
This and other instance variables in this file need to be used by either `@instance_var_name=blah` or `self.instance_var_name=blah`

```suggestion
            @current_error = nil
```